pkg/cover/backend/mach-o.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package backend
+
+import (
+	"debug/macho"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/google/syzkaller/pkg/host"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+func makeMachO(target *targets.Target, objDir, srcDir, buildDir string,
+	hostModules []host.KernelModule) (*Impl, error) {
+	// TODO: Linux/ELF supports module symbols. We should probably also do that
+	// for XNU/Mach-O. To maximize code re-use we already have a lot of the
+	// plumbing for module support. I think we mainly miss an equivalent to
+	// discoverModules and elfReadModuleCoverPoints at the moment.
+	if len(hostModules) != 0 {
+		return nil, fmt.Errorf("module coverage not yet supported for Mach-O")
+	}
+
+	return makeDWARF(target, objDir, srcDir, buildDir,
+		[]*Module{
+			// A dummy module representing the kernel itself.
+			{
+				Name: "",
+				Path: filepath.Join(objDir, target.KernelObject),
+				Addr: 0,
+			},
+		},
+		&containerFns{
+			readSymbols:           machoReadSymbols,
+			readCoverPoints:       machoReadCoverPoints,
+			readModuleCoverPoints: machoReadModuleCoverPoints,
+			readTextRanges:        machoReadTextRanges,
+		},
+	)
+}
+
+func machoReadSymbols(module *Module) ([]*Symbol, *symbolInfo, error) {
+	file, err := macho.Open(module.Path)
+	if err != nil {
+		return nil, nil, err
+	}
+	text := file.Section("__text")
+	if text == nil {
+		return nil, nil, fmt.Errorf("no __text section in the object file")
+	}
+	if file.Symtab == nil {
+		return nil, nil, fmt.Errorf("failed to read Mach-O symbols")
+	}
+	var symbols []*Symbol
+	info := &symbolInfo{
+		textAddr:    text.Addr,
+		traceCmp:    make(map[uint64]bool),
+		tracePCIdx:  make(map[int]bool),
+		traceCmpIdx: make(map[int]bool),
+	}
+
+	// We don't get symbol lengths or symbol ends in Mach-O symbols. So we
+	// guesstimate them by taking the next symbols beginning -1. That only
+	// works after we have sorted them.
+	sort.Slice(file.Symtab.Syms, func(i, j int) bool {
+		return file.Symtab.Syms[i].Value < file.Symtab.Syms[j].Value
+	})
+
+	for i, symb := range file.Symtab.Syms {
+		// Mach-Os doesn't contain the Symbol size like in ELF
+		symbEnd := text.Addr + text.Size
+		if i < len(file.Symtab.Syms)-1 {
+			symbEnd = file.Symtab.Syms[i+1].Value - 1
+		}
+
+		text := symb.Value >= text.Addr && symbEnd <= text.Addr+text.Size
+		if text {
```

Comment:
I assume ___sanitizer_cov_trace_ is also in the text section?
If yes, we can reduce nesting with `if !text { continue }`

Reply:
I don't quite get this one.  Do you mean `__sanitizer_cov_trace_cmp*`? If so then yes I guess they also must be in the text section. Later in this function we do this:
```
			if symb.Name == "___sanitizer_cov_trace_pc_guard" {
				info.tracePCIdx[i] = true
				if text {
					info.tracePC = symb.Value
				}
			} else {
				info.traceCmpIdx[i] = true
				if text {
					info.traceCmp[symb.Value] = true
				}
			}
```

I don't understand why we even record info.tracePCIdx and info.traceCmpIdx. If we really don't need these in the `!text` case, I'd agree that we can just continue right away and remove any further check. Do you have more context here?

Notice there is equivalent code in master for elf right now (again hard to reasonably de duplicate in this branch due to  debug/macho/#Symbol vs debug/elf/#Symbol):
https://github.com/google/syzkaller/blob/master/pkg/cover/backend/elf.go#L233

Reply:
> I don't understand why we even record info.tracePCIdx and info.traceCmpIdx. If we really don't need these in the !text case, I'd agree that we can just continue right away and remove any further check

Oh, right, it's more tricky.
So we need tracePCIdx/traceCmpIdx only for modules.
For the main kernel object, we only need the actual addresses in .text:
```
info.tracePC = symb.Value
info.traceCmp[symb.Value] = true
```
But modules don't contain these symbols in .text, instead they have some relocation symbols with these names I think. So we record indices of these relocation symbols to find calls of these symbols later.

Since for macho we don't support modules, and I don't even know if relocations are present and work the same way, we can simplify this to something along the following lines:
```
	for i, symb := range allSymbols {
		if symb.Value < text.Addr || symb.Value+symb.Size >= text.Addr+text.Size {
			continue
		}
		start := symb.Value + module.Addr
		symbols = append(symbols, &Symbol{
			Module: module,
			ObjectUnit: ObjectUnit{
				Name: symb.Name,
			},
			Start: start,
			End:   start + symb.Size,
		})
		if strings.HasPrefix(symb.Name, "__sanitizer_cov_trace_") {
			if symb.Name == "__sanitizer_cov_trace_pc" {
				info.tracePC = symb.Value
			} else {
				info.traceCmp[symb.Value] = true
			}
		}
	}
```

Reply:
Let's keep it as is.

Mach-O has relocations, though I don't yet know too much about them. I purposefully kept the rest of the code aligned with the ELF code and module support. I hope to add that in eventually and I think we probably aren't too far from it with this PR. IMO this change would be a minor win and potentially cause confusion down the line.